crates/iroha_core/src/state.rs

```diff
@@ -1618,61 +1547,172 @@ impl StateTransaction<'_, '_> {
         world.apply();
     }
 
-    fn process_executable(&mut self, executable: &Executable, authority: AccountId) -> Result<()> {
-        match executable {
-            Executable::Instructions(instructions) => {
-                self.process_instructions(instructions.iter().cloned(), &authority)
+    /// Execute a call-trigger. This function will be deprecated in #5147.
+    pub(crate) fn execute_called_trigger(
+        &mut self,
+        id: &TriggerId,
+        event: ExecuteTriggerEvent,
+    ) -> Result<(), TransactionRejectionReason> {
+        let (authority, executable) = {
+            let action = self
+                .world
+                .triggers
+                .by_call_triggers()
+                .get(event.trigger_id())
+                .ok_or_else(|| FindError::Trigger(id.clone()))
+                .map_err(Error::from)
+                .map_err(ValidationFail::from)?;
+            assert!(
+                !action.repeats.is_depleted(),
+                "orphaned trigger was not removed"
+            );
+
+            (action.authority().clone(), action.executable().clone())
+        };
+        self.world.external_event_buf.push(event.clone().into());
+        self.execute_trigger(id, &authority, &executable, event.into())?;
+        self.world.triggers.decrease_repeats([id].into_iter());
+
+        Ok(())
+    }
+
+    /// Perform a depth-first traversal of the trigger execution path.
+    pub(crate) fn execute_data_triggers_dfs(&mut self) -> Result<(), TransactionRejectionReason> {
+        let mut stack: Vec<(DataEvent, TriggerId, u8)> = self
+            .capture_data_events()
+            .into_iter()
+            // Preserve the order of the matched triggers
+            .rev()
+            .map(|(e, t)| (e, t, 1))
+            .collect();
+
+        while let Some((event, trg_id, depth)) = stack.pop() {
+            let max_depth = self.world.parameters.smart_contract.execution_depth;
```

Comment:
I'm debating whether to move this parameter retrieval outside the `while` loop.

If I do, any updates to the parameter won't take effect during the data-trigger chain -- the value stays fixed throughout. That's a bit odd, but it reliably prevents infinite loops.

If I leave it inside the loop, the parameter admin can raise the allowed depth mid-chain, which makes sense in principle -- but it also opens the door to infinite loops.

Reply:
I think it is consistent with the current behavior of adjusting trigger's repeats during its execution.

Reply:
I'd leave it as is. I think extending execution during execution is the responsibility of the parameter admin

Reply:
In any case, I think it is important to pin this behaviour as a test: create a trigger which recurses and increases the depth limit _to a certain degree_, then resets the parameter and completes.